c/certificateService.c

```diff
@@ -118,36 +119,80 @@ static void respondWithBadRequest(HttpResponse *response) {
 
     jsonStart(p);
     {
-      jsonAddString(p, "error", "The length of the certificate is longer than 4096 bytes");
+      jsonAddString(p, "error", value);
     }
     jsonEnd(p);
 
     finishResponse(response);
 }
 
+
 static int serveMappingService(HttpService *service, HttpResponse *response)
 {
   HttpRequest *request = response->request;
 
   if (!strcmp(request->method, methodPOST))
   {
-    RUsermapParamList *userMapCertificateStructure 
-      = (RUsermapParamList*)safeMalloc31(sizeof(RUsermapParamList),"RUsermapParamList");
-    memset(userMapCertificateStructure, 0, sizeof(RUsermapParamList));
-
-    if(request->contentLength > sizeof(userMapCertificateStructure->certificate) || request->contentLength < 0) {
-      respondWithBadRequest(response);
-      return 0;
+    RUsermapParamList userMapStructure;
+
+    memset(&userMapStructure, 0, sizeof(RUsermapParamList));
+
+    char translatedURL[strlen(request->uri) +1];
+    memcpy(translatedURL, request->uri, strlen(request->uri));
+    a2e(translatedURL, sizeof(translatedURL) - 1);
+    char *found = strstr(translatedURL,"x509");
+
+    if(found != NULL) {
+    //  Certificate to user mapping
+        if(request->contentLength > sizeof(userMapStructure.certificate) || request->contentLength < 0) {
+          respondWithBadRequest(response, "The length of the certificate is longer than 4096 bytes");
+          return 0;
+        }
+
+        userMapStructure.certificateLength = request->contentLength;
+        memset(userMapStructure.certificate, 0, request->contentLength);
+        memcpy(userMapStructure.certificate, request->contentBody, request->contentLength);
+
+        userMapStructure.functionCode = MAP_CERTIFICATE_TO_USERNAME;
+    } else {
+    //    Distinguished ID to user mapping
+        char *nb = copyStringToNative(request->slh, request->contentBody, strlen(request->contentBody));
+        char errB[2048];
+        Json *body = jsonParseUnterminatedString(request->slh, nb, strlen(nb), errB, sizeof(errB));
+        JsonObject *jsonObject = jsonAsObject(body);
+
+        char *distinguishedId = jsonObjectGetString(jsonObject, "dn");
+        if(distinguishedId == NULL || strlen(distinguishedId) == 0) {
+            respondWithBadRequest(response, "dn field not included in request body");
+            return 0;
+        } else if(strlen(distinguishedId) > sizeof(userMapStructure.distinguishedName)){
+            respondWithBadRequest(response, "The length of the distinguished name is more than 246 bytes");
+            return 0;
+        }
+        e2a(distinguishedId, strlen(distinguishedId));
```

Comment:
I think it's better to `e2a` the result buffer, not the source buffer. Who knows how else it's used, you don't control it.

Reply:
converting the result buffer now